### DISCUSSION TITLE
Trouble with validating merkle proof

### DISCUSSION THREAD

**Discussion by june-cho-88** [gcash/bchd#485] [Q&A]
I just picked a random transaction to validate which is _**94c8536b564c166286d850f9a93835703463e92403d8267e0bb81117242a808c**_.

My GetMerkleProofResponse gave these following data:
- Merkle Root
> _0c9b4118bfddce1211217c414cf5bf7be90365042ae7581c4106f11dbc1c8caa_
- 11 Merkle Proof Hashes
> [0]_12d129303667035e96b34d69fd4bc701ed9b01a20c2c2c8978624505b4f23f9c_
> [1]_3ad3d86e85db9a526651613e305d95df459a63b213f0aa73ad93cc3abb31bc7e_
> [2]_961fad10818b0063f01bd60700c673a6ad3ed955abae509adbd6147506a44028_
> [3]_32a5c9fcb1bd6e36523f24e21c9100d5081f90a3f17301ae2f5ca64db071d0d7_
> [4]_20eacb642554683c12f5bb6fcb00858ce7ce234593f4a4a96f2178bd88d3a65d_
> [5]_b508068ef823ef6285c086c24f22ae06da6816cb1e09915c62e8c1398f6c0ba8_
> [6]_d1de2349af80edda1cc12f5641160884042e11adc225e1e636266264ded4598c_
> [7]_94c8536b564c166286d850f9a93835703463e92403d8267e0bb81117242a808c_
> [8]_f34118013afb1179d4a470e849c748e79459bcf8eeea72db7a990284dc3e56ad_
> [9]_e590dd31a2abc736bd6085527c9ec19da0285427e0b3ac0040ef44d3cd0c8786_
> [10]_0452de19385dc9556526e0e14201cf4a124844f790beba7a6eb673187cfcd467_
- Flags
> _0b101010111011011000000010_ (0xABB602)

When I'm trying to build a partial merkle tree based on [this documentation](https://bchinfo.org/en/developer-reference#parsing-a-merkleblock-message), it doesn't match with full merkle tree.
For example, first of merkle proof hashes, [0]_12d129303667035e96b34d69fd4bc701ed9b01a20c2c2c8978624505b4f23f9c_, should be on a left side of **depth 2** according to full merkle tree, but the second flag bit(0) says it is on a left of a root, **depth 1**.

It seems like there is nothing wrong with bchd or transaction itself.
Did I misunderstand the documentation or something?

**Comment by cpacia** [comment]
Let me see if I understand it `0b101010111011011000000010`

- The first bit (1) means the left hash (we'll call this X) at depth 1 needs to be computed 
- The second bit (0) means the left hash for X, at depth 2, is the first hash in the provided list of hashes (0x12d12930)

That looks right to me. Am I reading it wrong?

**Reply by june-cho-88** [reply]
![image](https://bchinfo.org/img/dev/animated-en-merkleblock-parsing.gif)

I thought first bit represents the merkle root that needs to be computed later.
I try to use your approach for the next proof hash, but I'm stuck again.

Here is a complete partial merkle tree that gave me a correct merkle root hash.

Depth 02, Left = proofHashes[0]
Depth 03, Left = proofHashes[1]
Depth 04, Left = proofHashes[2]
Depth 05, Left = proofHashes[3]
Depth 07, Left = proofHashes[4]
Depth 09, Left = proofHashes[5]
Depth 10, Left = proofHashes[6]
Depth 10, Right = proofHashes[7]
Depth 08, Right = proofHashes[8]
Depth 06, Right = proofHashes[9]
Depth 01, Right = proofHashes[10]

**Reply by june-cho-88** [reply]
Here is a partial merkle tree I built from a full merkle tree with full transactions.

![image](https://user-images.githubusercontent.com/29689799/130393670-9a1932d3-d795-43a4-9a2e-7dd043630ba2.jpeg)


At first, I was following a description from bchd.
> Binary representing the location of the matching transaction in the full merkle tree, starting with the root (1) at position/level 0, where 1 corresponds to a left branch and 01 is a right branch.

So I tried to find a sequence of 4 same bits in flag bits since there is left-sided pattern that appeared from depth 2 to depth 5, but I couldn't.
What am I missing?
